File: Swift/Leetcode.playground/Sources/24_DP/LC_337_HouseRobberIII.swift
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init() { self.val = 0; self.left = nil; self.right = nil; }
 *     public init(_ val: Int) { self.val = val; self.left = nil; self.right = nil; }
 *     public init(_ val: Int, _ left: TreeNode?, _ right: TreeNode?) {
 *         self.val = val
 *         self.left = left
 *         self.right = right
 *     }
 * }
 */
class Solution {
    func rob(_ root: TreeNode?) -> Int {
        let res = robTree(root)
        return max(res[0], res[1])
    }
    
    private func robTree(_ node: TreeNode?) -> [Int] {
        guard let node = node else {
            return [0, 0]
        }
        let robLeft = robTree(node.left)
        let robRight = robTree(node.right)
        let max_1 = max(robLeft[0], robLeft[1]) + max(robRight[0], robRight[1])
        let max_2 = node.val + robLeft[0] + robRight[0]
        return [max_1, max_2]
    }
}

File: Java/24_DP/LC_337_HouseRobberIII.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
    public int rob(TreeNode root) {
        int[] res = robTree(root);
        return Math.max(res[0], res[1]);
    }

    private int[] robTree(TreeNode node) {
        if (node == null) {
            return new int[] { 0, 0 };
        }

        int[] robLeft = robTree(node.left);
        int[] robRight = robTree(node.right);

        int max_1 = Math.max(robLeft[0], robLeft[1]) + Math.max(robRight[0], robRight[1]);
        int max_2 = node.val + robLeft[0] + robRight[0];

        return new int[] { max_1, max_2 };
    }
}

public class LC_337_HouseRobberIII {

}
